cmd/openshift-install/gather.go

```diff
@@ -142,11 +143,17 @@ func runGatherBootstrapCmd(directory string) error {
 func logGatherBootstrap(bootstrap string, port int, masters []string, directory string) error {
 	logrus.Info("Pulling debug logs from the bootstrap machine")
 	client, err := ssh.NewClient("core", net.JoinHostPort(bootstrap, strconv.Itoa(port)), gatherBootstrapOpts.sshKeys)
-	if err != nil && strings.Contains(err.Error(), "ssh: handshake failed: ssh: unable to authenticate") {
-		return errors.Wrap(err, "failed to create SSH client, ensure the private key is added to your authentication agent (ssh-agent) or specified with the --key parameter")
-	} else if err != nil {
+	if err != nil {
+		if strings.Contains(err.Error(), "ssh: handshake failed: ssh: unable to authenticate") {
+			return errors.Wrap(err, "failed to create SSH client, ensure the private key is added to your authentication agent (ssh-agent) or specified with the --key parameter")
+		} else if errors.Is(err, syscall.ECONNREFUSED) {
+			return errors.Wrap(err, "failed to connect to the bootstrap machine")
+		} else if len(gatherBootstrapOpts.sshKeys) == 0 {
+			return errors.Wrap(err, "failed to create SSH client, ensure the proper ssh key is in your keyring or specify with --key")
+		}
```

Comment:
1. I think it doesn't have to be if else if, it can be written as blocks of if that return.

2. https://github.com/golang/crypto/blob/master/ssh/client_auth.go#L75 , https://github.com/golang/crypto/blob/master/ssh/client.go#L85 is where the authentication error is coming from and it doesn't look like it supports Unwrap so strings contains will have to do for now.

3. for unauthenticated error there are 2 cases,
- the env was set and https://github.com/openshift/installer/blob/master/pkg/gather/ssh/agent.go#L19
here maybe we can say failed to authenticate with the pre-existing ssh agent, add key to agent
- the keys were used
here we can say failed to authenticate using keys, add you should provide keys using --keys

or something like that.

4. len(gatherBootstrapOpts.sshKeys) == 0
I don't know what this case helps us with.

5. https://github.com/openshift/installer/blob/master/pkg/gather/ssh/agent.go#L19
uses ssh agent when no keys are provided, but that doesn't match the doc in https://github.com/openshift/installer/blob/master/docs/user/troubleshootingbootstrap.md#authenticating-to-bootstrap-host-for-ipi
also https://github.com/openshift/installer/commit/b91707efac5958d457c458d6af99d012dd22e12a#diff-be4790dcfa83ca14e7942b4c3b50c5f9 made it so that there is always one key passed to this list making it this ssh agent not used at all.. that would be bad.. maybe we fix that too, drop the `len(keys) == 0`

Reply:
> len(gatherBootstrapOpts.sshKeys) == 0
I don't know what this case helps us with

I added that case in 00307d19e6e42d533a3f4d99d31011a380cb8483 and I think it is a good idea to remove it in this PR.

Reply:
> for unauthenticated error there are 2 cases,

Determining which case happened seems to be a little hard based off of only the error message. https://github.com/openshift/installer/blob/master/pkg/gather/ssh/agent.go#L19 checks if the env is set and if it fails to dial, it goes with the other way of pulling the ssh keys from local. Any error message returned after that is in a general way and does not give details as to what case happened. I may be wrong.

The way I see it, I need to check if the env is set and if it dials right again here https://github.com/openshift/installer/blob/e2367c4f2b868cd5968fd3859dae2aa00aa7bcc4/cmd/openshift-install/gather.go#L145 which I think is repetitive and might take up time. Let me know if this is a viable option or if there is a better way.

Reply:
the function that has the information what was used to create the SSH Agent for authentication is the `ssh.NewClient` and i think we should look at how that function that provide us the context on which one was used so that we can decide the next steps when we fail to authenticate.

maybe we try that?